push_notifs: Add endpoint for sending a test notification.

mateuszmandera opened the pull request:
Still missing more complete tests, and I imagine we will want to add some error handling with informative error responses to make the endpoint useful?

Will fix #23997 when complete.

Adds two endpoints:
1. Remote server endpoint `users/me/test_notify`, to which the mobile app will be able to make a request.
2. This will cause the remote server to request the bouncer to send the notification to the device at `/api/v1/remotes/push/test_notify`.

A server like the bouncer itself skips step 2 of course and send the notification directly.

mateuszmandera commented on zerver/views/push_notifications.py:
```diff
@@ -66,3 +67,19 @@ def remove_android_reg_id(
     return json_success(request)
+
+
+@human_users_only
+@has_request_variables
+def send_test_push_notification_api(
+    request: HttpRequest, user_profile: UserProfile, token: str = REQ()
+) -> HttpResponse:
+    # TODO: Should we split this endpoint by token kind (apple/android)?
+    # Should we do any extra validation like the other endpoints above?
+    try:
+        device = PushDeviceToken.objects.get(token=token, user=user_profile)
+    except PushDeviceToken.DoesNotExist:
+        raise JsonableError(_("Token does not exist"))
```
I'm leaning towards the answer that we don't need to separate endpoints here or `kind`-based validation, because unlike the registration/deregistration endpoints above, we don't create some kind of db record that we would risk having malformed. But perhaps splitting would allow more helpful error responses for notification debugging? Not sure

timabbott replied:
I don't see much reason to split the endpoint by `kind`; though as noted in https://chat.zulip.org/#narrow/stream/378-api-design/topic/.2323997.20Endpoint.20for.20test.20notification/near/1638499, we might want have the `token` parameter be optional to support sending to all devices.

mateuszmandera replied:
Okay, removed the TODO and the "send to all devices" functionality is now implemented too.

mateuszmandera commented on zerver/lib/push_notifications.py:
```diff
@@ -1190,3 +1190,53 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
+    payload = base_payload
+    payload["event"] = "test-by-device-token"
+
+    if device.kind == RemotePushDeviceToken.APNS:
+        apns_data = {
+            "alert": {
+                "title": "Test notification",
+                "body": "This is a test notification from <realm URL>.",
+            },
+            "sound": "default",
+            "custom": {"zulip": payload},
+        }
+        send_apple_push_notification(user_identity, [device], apns_data, remote=remote)
+    else:
+        assert device.kind == RemotePushDeviceToken.GCM
+        payload["time"] = datetime_to_timestamp(timezone_now())
+        gcm_options = {"priority": "high"}
+        send_android_push_notification(user_identity, [device], payload, gcm_options, remote=remote)
```
This could use more eyes on whether the payloads are being formed correctly. I aimed to do it the way `get_message_payload_apns` and `get_message_payload_gcm`, combined with the input from https://github.com/zulip/zulip/issues/23997

mateuszmandera commented on zilencer/views.py:
```diff
@@ -281,6 +282,51 @@ def delete_duplicate_registrations(
+
+
+@has_request_variables
+def remote_server_send_test_notification(
```
To be clear, this is needed as a new bouncer endpoint, because the regular "send notifications" bouncer endpoint doesn't support the kind of request "send notification to this specific token" - and instead sends to all devices registered for the given user.

PIG208 replied:
For automatic payload validation with Pydantic, you can check out zerver/lib/typed_endpoint.py. This will require you to rewrite `payload` as a keyword-only argument, and annotate it as `payload: WebhookPayload[TestNotificationPayloadDict]`.

You might need to make adjustments to the `TypedDict` (refer to the Pydantic documentation [here](https://docs.pydantic.dev/latest/usage/types/types_fields/#typeddict)), or better, define it as a subclass of [`BaseModel`](https://docs.pydantic.dev/latest/usage/models/) for more straightforward `check_dict_only` equivalent support.
```py
from pydantic import BaseModel, ConfigDict

class TestNotificationPayload(BaseModel):
    server_uuid: str
    user_uuid: str
    user_id: int
    token: str
    token_kind: int
    base_payload: Dict[str, Any]

    model_config = ConfigDict(extra='forbid')
```

So the view will be:
```py
@typed_endpoint
def remote_server_send_test_notification(
    request: HttpRequest,
    server: RemoteZulipServer,
    *,
    # Equivalent to Annotated[Json[TestNotificationPayload], ApiParamConfig(argument_type_is_body=True)]
    # which, parses the request body as JSON and validates the parsed data using TestNotificationPayload
    payload: WebhookPayload[TestNotificationPayload],
) -> HttpResponse:
    ...
```

mateuszmandera replied:
Ah, that helps, the endpoint was pretty hastily and roughly typed for now - thanks!

timabbott replied:
We probably don't need a `server_uuid` parameter in that type -- that value should come from the `RemoteZulipServer` object. (I'm worried about security bugs where we check a different value than what we use).

mateuszmandera replied:
@PIG208 Do you think we should rename `WebhookPayload` to something general, like `JsonBodyPayload` or such? If you look a bit lower down in the file, the already existing `remote_server_notify_push` already uses this "payload as json in request's body" form of input - so maybe we shouldn't keep that type's name webhook-specific?

PIG208 replied:
Agreed. That new name makes more sense.

mateuszmandera replied:
@PIG208 I pushed with the endpoints typed as you advised, and also a commit for the `WebhookPayload` rename as said above. Can you take a look if this looks good now?

PIG208 replied:
The refactoring looks good to me. Thanks!

timabbott replied:
Extracted this as https://github.com/zulip/zulip/pull/26889 so we can integrate it without waiting for the rest of this PR to complete.

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -1190,3 +1190,53 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
+        apns_data = {
+            "alert": {
+                "title": "Test notification",
+                "body": "This is a test notification from <realm URL>.",
```
These should be tagged for translation, assuming they are displayed to the user.

mateuszmandera commented:
This has now addressed the feedback above and implemented the "send to all devices" feature requested in https://chat.zulip.org/#narrow/stream/378-api-design/topic/.2323997.20Endpoint.20for.20test.20notification/near/1638499. This is done by sending an empty value as `token` to the `users/me/test_notify` endpoint. It will cause the server to  make a request for  a test notification from the bouncer per each device we have registered on the server (for this user).

mateuszmandera commented on zproject/urls.py:
```diff
@@ -380,6 +381,7 @@
         "users/me/apns_device_token", POST=add_apns_device_token, DELETE=remove_apns_device_token
     ),
     rest_path("users/me/android_gcm_reg_id", POST=add_android_reg_id, DELETE=remove_android_reg_id),
+    rest_path("users/me/test_notify", POST=send_test_push_notification_api),
```
Do we want `zulip.yaml` docs for this, or to skip them for now like these other push endpoints?

mateuszmandera replied:
Pushed some documentation

mateuszmandera commented on zerver/openapi/zulip.yaml:
```diff
@@ -9096,6 +9096,49 @@ paths:
+    post:
+      operationId: test-notify
+      summary: Sends a test notifications to the user's mobile device(s)
+      tags: ["users"]
```
not sure about this tag, but matches better than any other tag we currently use i suppose

timabbott replied:
I think this should probably be in a new "mobile" tag.

mateuszmandera replied:
Fixed

mateuszmandera commented:
Pushed again renaming
`/users/me/test_notify` -> `/mobile_push/test_notification` as discussed.
Also renamed the bouncer endpoint from `.../test_notify` to `.../test_notification` for consistency.

chrisbobbe commented on zerver/lib/push_notifications.py:
```diff
@@ -1188,3 +1188,55 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
+        send_apple_push_notification(user_identity, [device], apns_data, remote=remote)
+    else:
+        assert device.kind == RemotePushDeviceToken.GCM
+        payload["time"] = timezone_now().timestamp()
```
I think this should be changed to be an `int`, for consistency with message notifications. How about `datetime_to_timestamp(timezone_now())`? Looking back at this GitHub thread, I think that was used in an earlier revision.

mateuszmandera replied:
Oh yeah, thanks for catching that, I missed to update this part of the PR after that revision

chrisbobbe commented:
Thanks! One comment on the payload: https://github.com/zulip/zulip/pull/26697#discussion_r1347817978.

Also, for the case where the server doesn't recognize the device token, I wonder if there's a better client-facing error message than "Token does not exist." After all, as long as the mobile client is behaving normally, the value it sends will be a real token that does indeed exist. 🙂 How about "Device token not recognized."?

timabbott commented:
Would "Device not recognized" be better? We should make sure that error returns a `code`, since the client error handling probably wants to try re-registering the device in response to it rather than displaying a weird error to the user.

mateuszmandera commented:
@chrisbobbe @timabbott This adds a bunch of complexity, so I pushed a separate commit that improves the error responses in the "Invalid token" cases to have distinct code - OpenAPI doc included for these two new responses. It's very basic still, and I imagine we will significantly expand on it as we come up with concrete classes of issues that this endpoint should allow the client to successfully debug :thinking:

timabbott commented on zerver/views/push_notifications.py:
```diff
@@ -66,3 +67,25 @@ def remove_android_reg_id(
+@human_users_only
+@has_request_variables
+def send_test_push_notification_api(
+    request: HttpRequest, user_profile: UserProfile, token: str = REQ()
```
Hmm, would it make sense to do a `default=None` thing for `token`? I feel like that's how we usually handle optional parameters...

timabbott commented on zerver/tests/test_push_notifications.py:
```diff
@@ -139,6 +146,186 @@ def get_generic_payload(self, method: str = "register") -> Dict[str, Any]:
+        ) as mock_send_android_push_notification, time_machine.travel(
+            time_now, tick=False
+        ):
+            result = self.api_post(user, endpoint, {"token": ""}, subdomain="zulip")
```
Yeah, this is an explicit `""` string as a token, rather than just not passing the parameter; makes me think our `REQ` configuration is off.

timabbott commented:
Generally lgtm aside from the minor issues noted above. We should add an API changelog entry and document that in `zulip.yaml`; but otherwise I'm not aware of anything else that makes this WIP? 

(I think we can very readily add additional error codes in follow-ups and that's probably a good idea to manage review effort.

alexmv commented on zerver/lib/push_notifications.py:
```diff
@@ -1188,3 +1188,55 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
+        send_apple_push_notification(user_identity, [device], apns_data, remote=remote)
+    else:
+        assert device.kind == RemotePushDeviceToken.GCM
+        payload["time"] = datetime_to_timestamp(timezone_now())
```
I don't think this is up to date with fd6091ad17b5e9e9e786418e3e8053332c67c0f9 and the code in general.  `get_base_payload` (passed in as `base_payload` here) is going to contain a `time` key already, with the appropriate granularity.

mateuszmandera replied:
Hmm, `time` is inserted in `get_message_payload` currently, not `get_base_payload`. But possibly we should move that into `get_base_payload`? That'll also mean `time` will be added in the `get_remove_payload_apns` `get_remove_payload_gcm` codepaths, but probably that's fine and shouldn't cause compatibility issues with the apps?

alexmv replied:
Ah, you're right that I was confusing `get_message_payload` and `get_base_payload`.

As for if it'd be a problem for `remove` payloads, that [sounds like an #api design question](https://github.com/zulip/zulip/pull/26768#issuecomment-1741313824).

alexmv commented on zerver/views/push_notifications.py:
```diff
@@ -66,3 +67,25 @@ def remove_android_reg_id(
+    else:
+        devices = list(PushDeviceToken.objects.filter(user=user_profile))
+
+    for device in devices:
+        send_test_push_notification(user_profile, device)
```
`send_apple_push_notification` and `send_android_push_notification` both take lists of devices, and the content is the same within one type of device.  Is it possible to not iterate here, and instead let them do their own iteration (which will enable 7787fe3f4904060e43fec8f5b78aee63147a9220 to do its work)?

mateuszmandera replied:
Oh yeah, that's a good idea. Pushed a `[squash]` commit for that, can you take a look if it makes sense now? This will only matter on Cloud right now, since it doesn't use the bouncer. A server that uses the bouncer still has to make requests to it one by one per device - until we make this bouncer endpoint accept taking in a list of tokens from the server. Probably a useful improvement to make, but maybe as a follow-up issue after this is merged?

timabbott replied:
Yeah, that adjustment to the bouncer is probably reasonable as a follow-up.

alexmv replied:
Yeah, this was mostly about people using this on Zulip Cloud, not through the push bouncer.

The `[squash]` LGTM!

mateuszmandera commented:
@timabbott Yeah this is no longer WIP at this point I think, fixed the title. Pushed two `[squash]` commit for clarity for review what's changing right now - for the API change to use `None` instead of empty string, and for smart looping like @alexmv suggested in https://github.com/zulip/zulip/pull/26697#discussion_r1352848888

Responded in https://github.com/zulip/zulip/pull/26697#discussion_r1352843484, which still remains a tiny open item

timabbott commented:
Changes lgtm, I guess we're still waiting on https://github.com/zulip/zulip/pull/26697#discussion_r1352843484.

timabbott commented:
I think we can squash the squashable commits, and bump the API changelog.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9129,6 +9129,41 @@ paths:
+      tags: ["mobile"]
+      description: |
+        This endpoint allows a user to trigger a test push notification to their
+        selected mobile devices, or all their mobile devices.
```
This needs a changes entry and API feature level bump.

mateuszmandera commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
 
+**Feature level 217**
+
+* [`POST /mobile_push/test_notification`](/api/test-notify): Added new endpoint
+  to send a test push notification to a mobile device or devices.
```
I'm not adding a link to this in `rest-endpoints.md`, because that makes `zerver/openapi/test_curl_examples.py` try to validate the `zulip.yaml` example, but I don't think we can feasibly do that - since successful operation of this endpoint requires talking to the bouncer

timabbott replied:
Yeah that sounds reasonable for now; @laurynmm FYI, we should probably discuss in "api documentation", but probably we want to just add some additional filtering to that test.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -9142,6 +9142,47 @@ paths:
+        This endpoint allows a user to trigger a test push notification to their
+        selected mobile devices, or all their mobile devices.
+
+        **Changes**: New in Zulip 8.0 (feature level 217).
```
Added the `**Changes**` entry here.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9142,6 +9142,47 @@ paths:
+  /mobile_push/test_notification:
+    post:
+      operationId: test-notify
+      summary: Sends a test notifications to the user's mobile device(s)
```
"a test notification"

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9142,6 +9142,47 @@ paths:
+      summary: Sends a test notifications to the user's mobile device(s)
+      tags: ["mobile"]
+      description: |
+        This endpoint allows a user to trigger a test push notification to their
+        selected mobile devices, or all their mobile devices.
```
s/devices/device/ (for the first occurrence) — you can select one, or let it go to all, but not select multiple

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9142,6 +9142,47 @@ paths:
+        - name: token
+          in: query
+          description: |
+            The push token for the device to send the test notification to.
```
"The push token for the device to which to send the test notification."

Or some other rewording. The thing with the existing version is that its syntax is ambiguous — and at least for me, my first reading is that it means the *device* will send the test notification to this push token. Where what's meant is that an unspecified subject (implicitly the server) will send the test notification to the device associated with this push token.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -19571,6 +19612,33 @@ components:
+          description: |
+            ## Invalid push device token
+
+            A typical failed JSON response for when the push device token is invalid:
```
s/invalid/not recognized by the Zulip server/

Or "this Zulip server".

"Invalid" sounds like it'd be about the token being, like, malformed as an FCM or APNs device token, but that's not what this error is about.

gnprice reviewed (commented):
Thanks for building this! Belated comments below on the API docs.

I also tried this out just now on chat.zulip.org and I get a 500 response — [see chat](https://chat.zulip.org/#narrow/stream/378-api-design/topic/.2323997.20Endpoint.20for.20test.20notification/near/1668595). So there's presumably something to fix there.
